File: RB101_RB109_small_problems/easy_4/convert_a_string_to_a_signed_number.rb
DIGITS = {
  "0" => 0,
  "1" => 1,
  "2" => 2,
  "3" => 3,
  "4" => 4,
  "5" => 5,
  "6" => 6,
  "7" => 7,
  "8" => 8,
  "9" => 9,
  "A" => 10,
  "B" => 11,
  "C" => 12,
  "D" => 13,
  "E" => 14,
  "F" => 15
}

def string_to_integer(string)
  string_array = string.split("")
  num = 0
  exponent = 0
  until string_array.empty?
    num += (DIGITS[string_array.pop] * (10 ** exponent))
    exponent += 1
  end
  num
end

def string_to_signed_integer(string)
  sign = 1
  if string[0] == "+" || string[0] == "-"
    sign = string.slice!(0)
    sign == "-" ? sign = -1 : sign = 1
  end
  string_to_integer(string) * sign
end

p string_to_signed_integer('4321') == 4321
p string_to_signed_integer('-570') == -570
p string_to_signed_integer('+100') == 100

# Further Exploration
def string_to_signed_integer(string)
  sign = string.start_with?("-") ? -1 : 1
  number = string.start_with?("-", "+") ? string[1..-1] : string
  
  string_to_integer(number) * sign
end

File: JS210_fundamentals/basics/convert_a_string_to_a_signed_number.js


function stringToInteger(str) {
  const STRING_TO_DIGITS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
  let digits = str.split('').reverse().map(d => STRING_TO_DIGITS[d]);

  let num = 0;

  for (let index = 0; index < digits.length; index += 1) {
    num += digits[index] * Math.pow(10, index);
  }

  return num;
}

function stringToSignedInteger(str) {
  let sign = +1;

  if (str[0] === '-') {
    sign = -1;
    str = str.slice(1);
  } else if (str[0] == '+') {
    str = str.slice(1);
  }

  return sign * stringToInteger(str);
}

console.log(stringToSignedInteger('4321'));      // 4321
console.log(stringToSignedInteger('-570'));      // -570
console.log(stringToSignedInteger('+100'));      // 100

// LS Solution

function stringToSignedInteger(string) {
  switch (string[0]) {
    case '-': return -stringToInteger(string.slice(1));
    case '+': return stringToInteger(string.slice(1));
    default:  return stringToInteger(string);
  }
}